File: 42. Trapping Rain Water/trap.kt
//better solution , using space,



class solution{
 fun trap(height: IntArray): Int{

    val n=height.size
    var leftMax=IntArray(n)
    var rightMax=IntArray(n)

    leftMax[0]= height[0]
    for(i in 1 until n){
    leftMax[i]= maxOf(leftMax[i-1],height[i])
    }

    rightMax[n-1]=height[n-1]
    for(i in n-2 downTo 0){
    rightMax[i]= maxOf(rightMax[i+1],height[i])
    }

     var water =0
     for(i in 0 until ){
     water +=minOf(leftMax[i],rightMax[i])-height[i]
     }

     return water

 }
}


fum main(){
val n=readln().toInt()

val height= readln().trim().split("\\s+".toRegex).mapNotNull{it.toIntOrNull()}.toIntArray

val trapwater=solution()
val ans=trapwater.trap(height)

println(ans)
}

File: 42. Trapping Rain Water/trap.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution{
public:
int trap(vector<int>& height){

    int left=0;
    int right =heigth.size()-1;
    int water=0;

    int leftMax=0;
    int rightMax=0;

    while(left<right){

    if(height[left]<height[right]){

    if(height[left]>=leftMax){
    leftMax=height[left];
    }
    else{
    water+=leftMax-height[left];
    }
    left++;
    }
    else{
    if(height[right]>=rightMax){
    rightMax=height[right];
    }
    else{
    water+=rightMax-height[right];
    }
    right--;
    }

    }

    return water;


}
};

int main(){

int n;
cin>>n;

vector<int> height(n);
for(auto &i: height){
cin>>i;
}

Solution trapwater;
int ans=trapwater.trap(height);

cout<<ans;

}
